walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
-
+import java.util.Collections;
+/* Implementation requires that Collection of Events are all sorted in order of meeting start time */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> avaliableTimes = new ArrayList<>();
+
+    /* If the requested time span is longer than a whole day, return no avaliable meeting times */
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return avaliableTimes;
+    }
+
+    int blockStart = TimeRange.START_OF_DAY;
+    int blockStartOpt = TimeRange.START_OF_DAY;
+    Collection<String> requestAtten = request.getAttendees();
+    Collection<String> optionalAtten = request.getOptionalAttendees();
+
+    /* If there are only optional attendees requested, default set all optional attendeeds to be included */
+    boolean optionalIncluded = request.getAttendees().isEmpty();
+    
+    for (Event event : events) { 
+      /* Check if there are attendees in the event that are also requested attendees */
+      Collection<String> eventAtten = event.getAttendees();
+      boolean timeConflictFound = !Collections.disjoint(eventAtten, requestAtten);
+      boolean timeConflictFoundOpt = !Collections.disjoint(eventAtten, optionalAtten);
+      int eventStart = event.getWhen().start();
+      int eventEnd = event.getWhen().end();
+
+      if (!timeConflictFoundOpt && eventStart - blockStartOpt >= request.getDuration()) {
+        optionalIncluded = true;
+      }
+
```

Comment:
Nit: remove this new line so it's clear that the if/else if blocks are logically connected.